### ISSUE TITLE
Workaround for torch.sort()?

### ISSUE THREAD

**Issue by coodoo** [hughperkins/cltorch#62]
I tried to run following test script to make predictions with a opencl-trained model:

```lua
mlp:cl()
local outputs = mlp:forward(inputs[1]):exp()
local probs, idx = torch.sort( outputs, true )
```
but got this error message `torch.ClTensor does not implement the torch.sort()`, current workaround is to move the model back to CPU using `mlp:float()` but it quickly become rather tedious, just wondering if there's better way to do it? Maybe a pure OpenCL way of sorting?

**Comment by hughperkins** [comment]
Hey! So, looking at the cuda version, there are two implementations.  One using thrust, one without:

```
  if (sliceSize <= 2048) {
    // Fill `indices` (the values) with the
    // slice-relative index.
    THCudaTensor_fillSliceWithIndex(state, indices, dim);

    // We sort k/v pairs in-place; copy unsorted input to output
    THCudaTensor_copy(state, sorted, input);

    // Sort using our in-place k/v kernel that supports arbitrary
    // layout
    THCudaTensor_sortKeyValueInplace(state, sorted, indices, dim, order);
  } else {
    // Otherwise, fall back upon Thrust, which handles all other cases
    // (potentially slowly, with extra copies/memory allocations)
    sortViaThrust(state, sorted, indices, input, dim, (bool) order);
  }
```
https://github.com/torch/cutorch/blob/master/lib/THC/THCTensorSort.cu#L447-L462

thrust isn't available on opencl, so we'd need to implement it ourselves somehow, I mean the sort, but it looks like there is already a cuda kernel available for slices up to size 2048, that would plausibly be straigthforward to port to opencl.  To what extent would slice sizes up to 2048 meet your requirements?

**Comment by coodoo** [comment]
Haha, fortunately this is a sentiment analysis task which only has 2 final classes (happy or sad, 1 or 2), and each batch has 10 sentences resulting in `[torch.Tensor 10x2]` (does this mean sorting 20 elements?), if so, 2048 should be more than enough 😆

**Comment by hughperkins** [comment]
I implemented sort, but for now it only works if you sort along the first dimension :-P  I was hoping that would be enough, but re-reading your comment, seems you probably need it to sort along the second dimension?

**Comment by hughperkins** [comment]
(actually, I'm not even sure it works for outermost dim.  i think I need to do a bit more testing/debugging)

**Comment by coodoo** [comment]
Here's how the output looks like, then I sort it to get the final classification results by reading `idx[i][1]`.

```lua
local outputs = mlp:forward(inputs[1]):exp()
 0.9999  0.0001
 0.0137  0.9862
 0.0002  0.9998
 0.9999  0.0001
 0.0000  1.0000
 0.0059  0.9941
 1.0000  0.0000
 0.9984  0.0015
 0.9949  0.0051
 0.0019  0.9980
[torch.FloatTensor of size 10x2]

-- then sort the outputs to get final result by reading idx
local p, idx = torch.sort( outputs, true )
```

Two questions:

1. Is there other way to get the final classification results other than `exp()` then `sort()`?

2. Instead of moving the whole model back to CPU, could I just do `outputs:float()` to move the it (which is a lot smaller) to CPU then sort it?

**Comment by hughperkins** [comment]
> Instead of moving the whole model back to CPU, could I just do outputs:float() to move the it (which is a lot smaller) to CPU then sort it?

Yes :-)

**Comment by hughperkins** [comment]
> Is there other way to get the final classification results other than exp() then sort()

If you want top-5, probably not.  If you want top-1, then you can use `max`

**Comment by hughperkins** [comment]
(by the way, not sure that you need to do `:exp()`?  since, exp is monotonic, so if `exp(a) > exp(b)`, then `a > b` ?

**Comment by coodoo** [comment]
Haha, actually I'm curious about `exp` too, was just following coursera's code sample [here](https://github.com/torch/tutorials/blob/master/coursera_neural_nets/assignment2/solution_torch_nn/predict_next_word.lua#L34-L36).

Ps. I thought the purpose of `exp` is to make something negative become positive?

**Comment by hughperkins** [comment]
> Ps. I thought the purpose of exp is to make something negative become positive?

It has that effect, though I'm not sure that is the purpose.  `max` works on negative numbers too: it returns the most positive term.  So max of -2 and -1, is -1.  max of -2, 1, and -1, is 1

**Comment by coodoo** [comment]
Yep that is reasonable doubt, it's just I saw a couple of similar examples doing it that way (a combination of `exp` with `sort` so I thought that was *the* formal way of doing it, would really love to know if there's any other solution? 😁

**Comment by hughperkins** [comment]
I dunno.  I always use max :-D  eg, see https://github.com/hughperkins/pytorch/blob/master/examples/luamodel/torch_model.lua#L135

**Comment by coodoo** [comment]
hey that's nice, just verified `max` is indeed faster and easier than `exp` and `sort`, let alone everything could be done on GPU, thanks for pointing out the right direction!

Btw, `max` also rendered this issue irrelevant, please feel free to close it if you don't have the bandwidth to fix `sort` in opencl.

**Comment by hughperkins** [comment]
Ok great!

**Comment by hughperkins** [comment]
Note: `sort` is working now :-)
